File: idekCTF/ski/interpreter.py
import re
import sys
from dataclasses import dataclass
from functools import cache
from pathlib import Path


sys.setrecursionlimit(2_000_000)


class Term:
    __slots__ = ()


class _Const(Term):
    __slots__ = ('_name',)

    def __init__(self, n: str) -> None:
        self._name = n

    def __repr__(self) -> str:
        return self._name

    def __hash__(self) -> int:
        return hash(self._name)

    def __eq__(self, other: object) -> bool:
        return self is other


S = _Const('S')
K = _Const('K')
I = _Const('I')  # noqa: E741


@dataclass(frozen=True, slots=True)
class Var(Term):
    name: str

    def __repr__(self) -> str:
        return self.name


@dataclass(frozen=True, slots=True)
class App(Term):
    f: Term
    x: Term

    def __repr__(self) -> str:
        return f'({self.f} {self.x})'


def show(t: Term) -> str:
    if t is S:
        return 'S'
    if t is K:
        return 'K'
    if t is I:
        return 'I'
    if isinstance(t, Var):
        return t.name
    if isinstance(t, App):
        return f'({show(t.f)} {show(t.x)})'
    return '?'


TOKENS = [
    ('SPACE', r'[ \t\r\n]+'),
    ('LP', r'\('),
    ('RP', r'\)'),
    ('EQ', r'='),
    ('ID', r'[A-Za-z_][A-Za-z0-9_]*'),
]
Token = tuple[str, str, int]


def lex(src: str) -> list[Token]:
    pos = 0
    out: list[Token] = []

    rx = re.compile('|'.join(f'(?P<{n}>{r})' for n, r in TOKENS))
    while pos < len(src):
        m = rx.match(src, pos)
        if not m:
            frag = src[pos : pos + 20].replace('\n', ' ')
            raise SyntaxError(f'Lex error at {pos}: {frag!r}')

        typ, val = m.lastgroup, m.group()
        pos = m.end()
        if typ == 'SPACE':
            continue
        out.append((typ, val, pos))
    return out


class Parser:
    def __init__(self, toks: list[Token]) -> None:
        self.toks = toks
        self.i = 0

    def peek(self, *k: str) -> Token | None:
        if self.i >= len(self.toks):
            return None
        t, v, p = self.toks[self.i]
        if not k or t in k or v in k:
            return (t, v, p)
        return None

    def eat(self, *k: str) -> Token | None:
        t = self.peek(*k)
        if not t:
            want = ' or '.join(k) if k else '(any)'
            got = self.toks[self.i] if self.i < len(self.toks) else 'EOF'
            raise SyntaxError(f'Parse error: expected {want}, got {got}')
        self.i += 1
        return t

    def parse_program(self) -> list[tuple[str, ...]]:
        items: list[tuple[str, ...]] = []
        while self.i < len(self.toks):
            if self.peek('ID') and self.i + 1 < len(self.toks) and self.toks[self.i + 1][0] == 'EQ':
                name = self.eat('ID')[1]
                self.eat('EQ')
                items.append(('def', name, self.parse_term()))
            else:
                items.append(('term', self.parse_term()))
        return items

    def parse_term(self) -> 'Term':
        level: list[list[Term]] = [[]]

        while self.i < len(self.toks):
            if self.peek('LP'):
                self.eat('LP')
                level.append([])
                continue

            if self.peek('RP'):
                self.eat('RP')
                if len(level) == 1:
                    break
                finished = self._chain(level.pop())
                level[-1].append(finished)
                continue

            if self.peek('ID'):
                _, name, _ = self.eat('ID')
                if name == 'S':
                    term = S
                elif name == 'K':
                    term = K
                elif name == 'I':
                    term = I
                else:
                    term = Var(name)
                level[-1].append(term)
                continue

            break

        return self._chain(level.pop())

    @staticmethod
    def _chain(ts: list['Term']) -> 'Term':
        if not ts:
            raise SyntaxError
        res = ts[0]
        for a in ts[1:]:
            res = App(res, a)
        return res


def parse(src: str) -> list[tuple[str, ...]]:
    return Parser(lex(src)).parse_program()


def prelude(flag: str, max_bits: int = 128 * 8) -> dict[str, Term]:
    bits = []
    for b in flag.encode():
        bits.extend((b >> i) & 1 for i in range(7, -1, -1))
        if len(bits) >= max_bits:
            break
    bits.extend([0] * (max_bits - len(bits)))
    return {f'_F{i}': (K if bit else App(K, I)) for i, bit in enumerate(bits[:max_bits])}


def substitute(t: Term, env: dict[str, Term], cache: dict[Term, Term] | None = None) -> Term:
    if cache is None:
        cache = {}

    if t in cache:
        return cache[t]

    if t is S or t is K or t is I:
        cache[t] = t
        return t

    if isinstance(t, Var):
        res = substitute(env[t.name], env, cache) if t.name in env else t
        cache[t] = res
        return res

    if isinstance(t, App):
        f_new = substitute(t.f, env, cache)
        x_new = substitute(t.x, env, cache)
        res = t if (f_new is t.f and x_new is t.x) else App(f_new, x_new)
        cache[t] = res
        return res

    raise TypeError(t)


def _rebuild(head: Term, args: list[Term]) -> Term:
    for a in args:
        head = App(head, a)
    return head


def _flatten(t: Term) -> tuple[Term, list[Term]]:
    args: list[Term] = []
    while isinstance(t, App):
        args.append(t.x)
        t = t.f
    args.reverse()
    return t, args


def _step(t: Term) -> Term | None:
    head, args = _flatten(t)

    if head is I and args:
        return _rebuild(args[0], args[1:])

    if head is K and len(args) >= 2:
        return _rebuild(args[0], args[2:])

    if head is S and len(args) >= 3:
        f, g, x, *rest = args
        return _rebuild(App(App(f, x), App(g, x)), rest)

    if isinstance(head, App):
        h2 = _step(head)
        if h2 is not None:
            return _rebuild(h2, args)

    for i, a in enumerate(args):
        if isinstance(a, App):
            a2 = _step(a)
            if a2 is not None:
                args = args.copy()
                args[i] = a2
                return _rebuild(head, args)

    return None


@cache
def normal_form(t: Term) -> tuple[Term, int]:
    steps = 0
    cur = t
    while True:
        nxt = _step(cur)
        if nxt is None:
            return cur, steps
        cur = nxt
        steps += 1


def main() -> None:
    with Path('./program.txt').open() as f:
        src = f.read()

    print('₍^. .^₎⟆')
    flag = input().strip()

    print('parsing')
    items = parse(src)
    env = prelude(flag)
    last_term = None

    for it in items:
        if it[0] == 'def':
            _, name, term_ast = it
            env[name] = substitute(term_ast, env)
        else:
            _, term_ast = it
            last_term = substitute(term_ast, env)

    print('reducing to normal form')
    nf, steps = normal_form(last_term)
    print(f'reduced in {steps} steps!')
    print(f'- nf: {show(nf)}')

    if nf is K:
        print('- flag: correct')
    elif isinstance(nf, App) and nf.f is K and nf.x is I:
        print('- flag: incorrect')
    else:
        print('- flag: unknown status (did you mess something up?)')


if __name__ == '__main__':
    main()

File: idekCTF/ski/interpreter.cpp
/// `clang++ -std=c++20 -O3 -pipe -static -march=native -flto interpreter.cpp -o interpreter`
/// `g++ -std=c++20 -O3 -pipe -static -march=native -flto interpreter.cpp -o interpreter`
/// msvc is supported, but you're on your own :)
#include <algorithm>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <span>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

struct Term;
using TermPtr = Term*;
using CTermPtr = const Term*;

enum struct Tag : uint8_t {
    S,
    K,
    I,
    Var,
    App
};

struct alignas(8) Term {
    union {
        struct {
            TermPtr f, x;
        };
        uint32_t name;
    };
};

namespace detail {
    inline constexpr uintptr_t TAG_MASK = 0b111; // (<= 8 tags)
    inline constexpr uintptr_t PTR_MASK = ~TAG_MASK;

    inline TermPtr tag_ptr(Tag t, TermPtr p) noexcept {
        return reinterpret_cast<TermPtr>((reinterpret_cast<uintptr_t>(p) & PTR_MASK) | static_cast<uintptr_t>(t));
    }
    inline Tag get_tag(CTermPtr p) noexcept {
        return static_cast<Tag>(reinterpret_cast<uintptr_t>(p) & TAG_MASK);
    }
    inline TermPtr untag_ptr(TermPtr p) noexcept {
        return reinterpret_cast<TermPtr>(reinterpret_cast<uintptr_t>(p) & PTR_MASK);
    }
} // namespace detail

class AllocationArena {
    static constexpr std::size_t CHUNK = 1 << 20; // 1 MiB

    struct Chunk {
        Chunk* next;
        char data[CHUNK - sizeof(Chunk*)];
    };

    Chunk* head_ = nullptr;
    char* cur_ = nullptr;
    char* end_ = nullptr;

    void new_chunk() {
        auto* c = static_cast<Chunk*>(std::malloc(sizeof(Chunk)));
        if (!c) {
            throw std::bad_alloc{};
        }
        c->next = head_;
        head_ = c;
        cur_ = c->data;
        end_ = c->data + sizeof(c->data);
    }

public:
    ~AllocationArena() {
        while (head_) {
            Chunk* nxt = head_->next;
            std::free(head_);
            head_ = nxt;
        }
    }

    template <class... A>
    [[nodiscard]] TermPtr make(const Tag tag, A... a) {
        constexpr std::size_t need = sizeof(Term);
        if (cur_ == nullptr || cur_ + need > end_) {
            new_chunk();
        }

        auto* raw = reinterpret_cast<TermPtr>(cur_);
        cur_ += need;

        if constexpr (sizeof...(A) == 2) {
            raw->f = std::get<0>(std::tuple<A...>(a...));
            raw->x = std::get<1>(std::tuple<A...>(a...));
        } else if constexpr (sizeof...(A) == 1) {
            raw->name = std::get<0>(std::tuple<A...>(a...));
        }
        return detail::tag_ptr(tag, raw);
    }
} arena;

std::unordered_map<std::string_view, uint32_t> id_of;
std::vector<std::string_view> name_of;

[[nodiscard]] uint32_t intern(const std::string_view s) {
    if (const auto it = id_of.find(s); it != id_of.end()) {
        return it->second;
    }

    const auto id = static_cast<uint32_t>(name_of.size());
    /// CHECK: we are sure that the string will still be in memory
    auto [it, _] = id_of.emplace(s, id_of.size());
    name_of.emplace_back(it->first);
    return id;
}

[[nodiscard]] TermPtr S() {
    static TermPtr p = arena.make(Tag::S);
    return p;
}

[[nodiscard]] TermPtr K() {
    static TermPtr p = arena.make(Tag::K);
    return p;
}

[[nodiscard]] TermPtr I() {
    static TermPtr p = arena.make(Tag::I);
    return p;
}

[[nodiscard]] TermPtr Var(const uint32_t id) {
    return arena.make(Tag::Var, id);
}

[[nodiscard]] TermPtr App(TermPtr f, TermPtr x) {
    return arena.make(Tag::App, f, x);
}

template <class Node>
auto& get_f(Node* p) {
    return detail::untag_ptr(p)->f;
}
template <class Node>
auto& get_x(Node* p) {
    return detail::untag_ptr(p)->x;
}
template <class Node>
auto& get_name(Node* p) {
    return detail::untag_ptr(p)->name;
}

struct Parser {
    const char *s, *p;
    explicit Parser(const std::string_view src): s(src.data()), p(s) { }
    explicit Parser(const char* src): s(src), p(s) { }

    void trim_whitespaces() {
        while (isspace(*p)) {
            ++p;
        }
    }

    [[nodiscard]] TermPtr parse_term() {
        trim_whitespaces();
        if (*p == '(') {
            ++p;
            TermPtr t = parse_expr();
            trim_whitespaces();
            if (*p != ')')
                err();
            ++p;
            return t;
        }

        if (isalpha(*p) || *p == '_') {
            const char* b = p++;
            while (isalnum(*p) || *p == '_') {
                ++p;
            }

            std::string_view name(b, static_cast<std::size_t>(p - b));
            if (name == "S") {
                return S();
            }
            if (name == "K") {
                return K();
            }
            if (name == "I") {
                return I();
            }
            return Var(intern(name));
        }

        err();
    }

    TermPtr parse_expr() {
        TermPtr r = parse_term();
        while (true) {
            trim_whitespaces();
            if (*p == '(' || isalpha(*p) || *p == '_') {
                r = App(r, parse_term());
            } else {
                break;
            }
        }
        return r;
    }

    [[noreturn]] static void err() {
        throw std::runtime_error("parse error");
    }
};

std::string show(TermPtr t) {
    switch (detail::get_tag(t)) {
    case Tag::S:
        return "S";
    case Tag::K:
        return "K";
    case Tag::I:
        return "I";
    case Tag::Var:
        return std::string(name_of[get_name(t)]);
    case Tag::App:
        return "(" + show(get_f(t)) + " " + show(get_x(t)) + ")";
    }
    return "?";
}

struct PtrHash {
    [[nodiscard]] size_t operator()(TermPtr p) const noexcept {
        return reinterpret_cast<uintptr_t>(p);
    }
};

struct PtrEq {
    [[nodiscard]] bool operator()(const CTermPtr a, const CTermPtr b) const noexcept {
        return a == b;
    }
};

[[nodiscard]] TermPtr substitute(TermPtr in_term, const std::unordered_map<uint32_t, TermPtr>& env,
                                 std::unordered_map<TermPtr, TermPtr, PtrHash, PtrEq>& memo) {
    if (const auto it = memo.find(in_term); it != memo.end()) {
        return it->second;
    }

    TermPtr term = in_term;
    switch (detail::get_tag(in_term)) {
    case Tag::Var: {
        if (const auto ie = env.find(get_name(in_term)); ie != env.end()) {
            term = substitute(ie->second, env, memo);
        }
        break;
    }
    case Tag::App: {
        TermPtr f = substitute(get_f(in_term), env, memo);
        TermPtr x = substitute(get_x(in_term), env, memo);
        term = (f == get_f(in_term) && x == get_x(in_term)) ? in_term : App(f, x);
        break;
    }
    default: {
        break;
    }
    }
    memo.emplace(in_term, term);
    return term;
}

[[nodiscard]] TermPtr rebuild(TermPtr head, const std::span<TermPtr> args) {
    for (TermPtr a : args) {
        head = App(head, a);
    }
    return head;
}

[[nodiscard]] TermPtr step(TermPtr term) {
    std::vector<TermPtr> args = {};
    TermPtr h = term;

    while (detail::get_tag(h) == Tag::App) {
        args.push_back(get_x(h));
        h = get_f(h);
    }

    std::ranges::reverse(args);
    size_t argc = args.size();
    auto get_arg = [&](std::size_t i) -> TermPtr& {
        return args[i];
    };

    if (h == I() && argc >= 1) {
        if (argc == 1) {
            return get_arg(0);
        }

        return rebuild(get_arg(0), std::span(&get_arg(1), argc - 1));
    }

    if (h == K() && argc >= 2) {
        if (argc == 2) {
            return get_arg(0);
        }

        return rebuild(get_arg(0), std::span(&get_arg(2), argc - 2));
    }

    if (h == S() && argc >= 3) {
        TermPtr f = get_arg(0);
        TermPtr g = get_arg(1);
        TermPtr x = get_arg(2);

        std::vector<TermPtr> rest;
        rest.reserve(argc - 3);
        for (size_t i = 3; i < argc; ++i) {
            rest.push_back(get_arg(i));
        }

        TermPtr head = App(App(f, x), App(g, x));
        return rebuild(head, rest);
    }

    if (detail::get_tag(h) == Tag::App) {
        if (TermPtr h2 = step(h)) {
            return rebuild(h2, {&get_arg(0), argc});
        }
    }

    for (size_t i = 0; i < argc; ++i) {
        if (auto arg = get_arg(i); detail::get_tag(arg) == Tag::App) {
            if (TermPtr a2 = step(arg)) {
                args[i] = a2;
                return rebuild(h, {&get_arg(0), argc});
            }
        }
    }
    return nullptr;
}

std::pair<TermPtr, int> normal_form(TermPtr t) {
    static std::unordered_map<TermPtr, std::pair<TermPtr, int>, PtrHash, PtrEq> cache;
    if (const auto it = cache.find(t); it != cache.end()) {
        return it->second;
    }

    int n = 0;
    TermPtr cur = t;
    while (TermPtr nx = step(cur)) {
        cur = nx;
        ++n;
    }
    return cache.emplace(t, std::make_pair(cur, n)).first->second;
}

std::unordered_map<uint32_t, TermPtr> prelude(const std::string& flag, size_t max_bits = 128 * 8) {
    std::vector<int> bits;
    bits.reserve(max_bits);
    for (const auto c : flag) {
        for (int i = 7; i >= 0 && bits.size() < max_bits; --i) {
            bits.push_back((c >> i) & 1);
        }
    }

    while (bits.size() < max_bits) {
        bits.push_back(0);
    }

    std::unordered_map<uint32_t, TermPtr> env;
    for (size_t i = 0; i < max_bits; ++i) {
        env[intern("_F" + std::to_string(i))] = bits[i] ? K() : App(K(), I());
    }
    return env;
}

int main() try {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::ifstream fin("program.txt");
    if (!fin) {
        std::cerr << "Could not open program.txt\n";
        return 1;
    }

    std::string src(std::istreambuf_iterator(fin), {});

    std::string flag;
    std::cout << "₍^. .^₎⟆" << std::endl;
    std::cin >> flag;

    std::cout << "parsing" << std::endl;
    std::vector<std::pair<bool, TermPtr>> prog;
    const char* s = src.c_str();
    const char* e = s + src.size();

    while (s < e) {
        while (s < e && isspace(*s)) {
            ++s;
        }

        if (s == e) {
            break;
        }

        const char* id_b = s;
        while (s < e && !isspace(*s) && *s != '=') {
            ++s;
        }

        while (s < e && isspace(*s)) {
            ++s;
        }

        if (s < e && *s == '=') {
            ++s;

            auto name = std::string_view{id_b, static_cast<std::size_t>(s - id_b)};
            TermPtr rhs = Parser(s).parse_expr();
            prog.emplace_back(true, App(Var(intern(name)), rhs));
            s = e;
        } else {
            prog.emplace_back(false, Parser(id_b).parse_expr());
            break;
        }
    }

    auto env = prelude(flag);
    TermPtr last = nullptr;
    for (auto& [is_def, node] : prog) {
        std::unordered_map<TermPtr, TermPtr, PtrHash, PtrEq> memo;
        if (is_def) {
            env[node->f->name] = substitute(node->x, env, memo);
        } else {
            last = substitute(node, env, memo);
        }
    }

    std::cout << "reducing to normal form" << std::endl;
    auto [nf, steps] = normal_form(last);
    std::cout << "reduced in " << steps << " steps!\n";
    std::cout << "- nf: " << show(nf) << "\n";

    if (nf == K()) {
        std::cout << "- flag: correct\n";
    } else if (detail::get_tag(nf) == Tag::App && get_f(nf) == K() && get_x(nf) == I()) {
        std::cout << "- flag: incorrect\n";
    } else {
        std::cout << "- flag: unknown status\n";
    }

    return 0;
} catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << '\n';
    return 1;
}
